File: src/main/java/com/github/kindrat/cassandra/client/ui/window/editor/tables/TablePanel.java
package com.github.kindrat.cassandra.client.ui.window.editor.tables;

import com.github.kindrat.cassandra.client.i18n.MessageByLocaleService;
import com.github.kindrat.cassandra.client.properties.UIProperties;
import com.github.kindrat.cassandra.client.ui.MainController;
import com.github.kindrat.cassandra.client.util.UIUtil;
import javafx.collections.ObservableList;
import javafx.geometry.Orientation;
import javafx.scene.control.SplitPane;
import javafx.scene.input.ContextMenuEvent;
import javafx.scene.layout.AnchorPane;
import org.springframework.beans.BeansException;
import org.springframework.beans.factory.BeanFactory;
import org.springframework.beans.factory.BeanFactoryAware;

import java.util.function.Consumer;

public class TablePanel extends AnchorPane implements BeanFactoryAware {
    private final TableListView tableListView;
    private final TableButtons buttons;
    private final TableListContext tableContext;
    private final MainController controller;

    public TablePanel(UIProperties uiProperties, MessageByLocaleService localeService, MainController controller) {
        this.controller = controller;
        setPrefHeight(uiProperties.getTablesPrefHeight());
        setPrefWidth(uiProperties.getTablesPrefWidth());

        buttons = new TableButtons();
        tableListView = new TableListView();
        tableListView.onMouseClick(buttons::enableButtons, buttons::disableButtons, this::tryLoadData);
        tableListView.setOnContextMenuRequested(this::onTableContextMenu);

        SplitPane splitPane = splitPane(uiProperties);
        UIUtil.fillParent(splitPane);
        getChildren().add(splitPane);

        tableContext = new TableListContext(localeService);
        tableContext.onDataAction(this::tryLoadData);
        tableContext.onDdlAction(this::tryLoadDDL);
        tableContext.onExportAction(this::tryExportData);
    }

    public void showTables(ObservableList<String> tables) {
        tableListView.showTables(tables);
    }

    public void clear() {
        tableContext.hide();
        tableListView.clear();
    }

    public void setNewValueListener(Consumer<String> newSelectedTableListener) {
        tableListView.onNewValueSelected((observable, oldValue, newValue) -> {
            if (newValue != null) {
                newSelectedTableListener.accept(newValue);
            }
            tableContext.hide();
        });
    }

    @Override
    public void setBeanFactory(BeanFactory beanFactory) throws BeansException {
        buttons.initActions(beanFactory);
    }

    private SplitPane splitPane(UIProperties uiProperties) {
        SplitPane splitPane = new SplitPane(tableListView, buttons);
        splitPane.setDividerPositions(uiProperties.getTablesDividerPosition());
        splitPane.setFocusTraversable(false);
        splitPane.setMaxHeight(40);
        splitPane.setPrefHeight(40);
        splitPane.setPrefWidth(160);
        splitPane.setScaleShape(false);
        splitPane.setOrientation(Orientation.VERTICAL);
        return splitPane;
    }

    private void tryLoadDDL() {
        tableListView.getSelectedTable().ifPresent(controller::showDDLForTable);
    }

    private void tryLoadData() {
        tableListView.getSelectedTable().ifPresent(controller::showDataForTable);
    }

    private void tryExportData() {
        tableListView.getSelectedTable().ifPresent(controller::exportDataForTable);
    }

    private void onTableContextMenu(ContextMenuEvent event) {
        tableListView.getSelectedTable()
                .ifPresent(table -> tableContext.show(this, event.getScreenX(), event.getScreenY()));
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/ui/window/editor/tables/TablePanel.kt
package com.github.kindrat.cassandra.client.ui.window.editor.tables

import com.github.kindrat.cassandra.client.i18n.MessageByLocaleService
import com.github.kindrat.cassandra.client.properties.UIProperties
import com.github.kindrat.cassandra.client.ui.MainController
import com.github.kindrat.cassandra.client.util.UIUtil.fillParent
import javafx.beans.value.ObservableValue
import javafx.collections.ObservableList
import javafx.event.EventHandler
import javafx.geometry.Orientation
import javafx.scene.control.SplitPane
import javafx.scene.input.ContextMenuEvent
import javafx.scene.layout.AnchorPane
import org.springframework.beans.BeansException
import org.springframework.beans.factory.BeanFactory
import org.springframework.beans.factory.BeanFactoryAware
import java.util.function.Consumer

class TablePanel(
        uiProperties: UIProperties,
        localeService: MessageByLocaleService,
        private val controller: MainController
) : AnchorPane(), BeanFactoryAware {

    companion object {
        const val ID = "RootTablePanel"
    }

    private val tableListView: TableListView
    private val buttons: TableButtons
    private val tableContext: TableListContext

    fun showTables(tables: ObservableList<String>) {
        tableListView.showTables(tables)
    }

    fun clear() {
        tableContext.hide()
        tableListView.clear()
    }

    fun setNewValueListener(newSelectedTableListener: Consumer<String?>) {
        tableListView.onNewValueSelected { _: ObservableValue<out String?>?, _: String?, newValue: String? ->
            if (newValue != null) {
                newSelectedTableListener.accept(newValue)
            }
            tableContext.hide()
        }
    }

    @Throws(BeansException::class)
    override fun setBeanFactory(beanFactory: BeanFactory) {
        buttons.initActions(beanFactory)
    }

    private fun splitPane(uiProperties: UIProperties): SplitPane {
        val splitPane = SplitPane(tableListView, buttons)
        splitPane.setDividerPositions(uiProperties.tablesDividerPosition)
        splitPane.isFocusTraversable = false
        splitPane.maxHeight = 40.0
        splitPane.prefHeight = 40.0
        splitPane.prefWidth = 160.0
        splitPane.isScaleShape = false
        splitPane.orientation = Orientation.VERTICAL
        return splitPane
    }

    private fun tryLoadDDL() {
        tableListView.selectedTable?.apply { controller.showDDLForTable(this) }
    }

    private fun tryLoadData() {
        tableListView.selectedTable?.apply { controller.showDataForTable(this) }
    }

    private fun tryExportData() {
        tableListView.selectedTable?.apply { controller.exportDataForTable(this) }
    }

    private fun onTableContextMenu(event: ContextMenuEvent) {
        tableListView.selectedTable?.apply { tableContext.show(this@TablePanel, event.screenX, event.screenY) }
    }

    init {
        id = ID
        prefHeight = uiProperties.tablesPrefHeight
        prefWidth = uiProperties.tablesPrefWidth
        buttons = TableButtons()
        tableListView = TableListView()
        tableListView.onMouseClick({ buttons.enableButtons() }, { buttons.disableButtons() }, { tryLoadData() })
        tableListView.onContextMenuRequested = EventHandler { onTableContextMenu(it) }
        val splitPane = splitPane(uiProperties)
        fillParent(splitPane)
        children.add(splitPane)
        tableContext = TableListContext(localeService)
        tableContext.onDataAction { tryLoadData() }
        tableContext.onDdlAction { tryLoadDDL() }
        tableContext.onExportAction { tryExportData() }
    }
}